dagstore.go

```diff
@@ -0,0 +1,602 @@
+	return d.queueTask(tsk, false)
+}
+
+func (d *DAGStore) control() {
+	defer d.wg.Done()
+
+	var tsk *Task
+	for {
+		select {
+		case tsk = <-d.taskCh:
+		case <-d.ctx.Done():
+			return
+		}
+
+		log.Infow("processing task", "op", tsk.Op, "shard", tsk.Shard.key)
+
+		if len(d.overflow) > 0 {
+			log.Info("moving an overflow item to task queue")
+			select {
+			case d.taskCh <- d.overflow[0]:
+				d.overflow = d.overflow[1:]
+			default:
+				log.Info("could not move overflow item into task queue")
+			}
+		}
+
+		s := tsk.Shard
+		s.Lock()
+
+		switch tsk.Op {
+		case OpShardRegister:
+			if s.state != ShardStateNew {
+				err := fmt.Errorf("%w: expected shard to be in 'new' state; was: %d", ErrShardInitializationFailed, s.state)
+				_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+				break
+			}
+			// queue a fetch.
+			_ = d.queueTask(&Task{Op: OpShardFetch, Shard: tsk.Shard}, true)
+
+		case OpShardFetch:
+			s.state = ShardStateFetching
+
+			go func(ctx context.Context, upgrader *mount.Upgrader) {
+				// ensure a copy is available locally.
+				reader, err := upgrader.Fetch(ctx)
+				if err != nil {
+					err = fmt.Errorf("failed to acquire reader of mount: %w", err)
+					_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+					return
+				}
+				_ = reader.Close()
+				_ = d.queueTask(&Task{Op: OpShardFetchDone, Shard: tsk.Shard}, true)
+			}(d.ctx, s.mount)
+
+		case OpShardFetchDone:
+			s.state = ShardStateFetched
+			if !s.indexed {
+				// shard isn't indexed yet, so let's index.
+				_ = d.queueTask(&Task{Op: OpShardIndex, Shard: tsk.Shard}, true)
+				break
+			}
+			// shard is indexed, we're ready to serve requests.
+			_ = d.queueTask(&Task{Op: OpShardMakeAvailable, Shard: tsk.Shard}, true)
+
+		case OpShardIndex:
+			s.state = ShardStateIndexing
+			go func(ctx context.Context, mnt mount.Mount) {
+				reader, err := mnt.Fetch(ctx)
+				if err != nil {
+					err = fmt.Errorf("failed to acquire reader of mount: %w", err)
+					_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+					return
+				}
+				defer reader.Close()
+
+				idx, err := loadIndex(reader)
+				if err != nil {
+					err = fmt.Errorf("failed to index shard: %w", err)
+					_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+					return
+				}
+				_ = d.queueTask(&Task{Op: OpShardIndexDone, Shard: tsk.Shard, Index: idx}, true)
+			}(d.ctx, s.mount)
+
+		case OpShardIndexDone:
+			err := d.indices.AddFullIndex(s.key, tsk.Index)
+			if err != nil {
+				err = fmt.Errorf("failed to add index for shard: %w", err)
+				_ = d.queueTask(&Task{Op: OpShardFail, Shard: tsk.Shard, Error: err}, true)
+				break
+			}
+			_ = d.queueTask(&Task{Op: OpShardMakeAvailable, Shard: tsk.Shard}, true)
+
+		case OpShardMakeAvailable:
+			if s.wRegister != nil {
+				res := ShardResult{Key: s.key}
+				go func(ch chan ShardResult) { ch <- res }(s.wRegister)
```

Comment:
Unbounded growth in goroutines here if client supplies a nil channel to RegisterShard or is slower than the dagstore at reading from the supplied channel

Reply:
An alternate API design that avoids the need for the dagstore to manage the lifetimes of these goroutines.

Current client pattern looks like this:

```
results := make(chan ShardResult, 100)

go func() {
	select {
		case res := <-results:
			// do something with the result
	}
}()

for _, key := range shards {
	dagstore.RegisterShard(key, mnt, results, opts)
}
```

Client MUST service the results channel otherwise the dagstore will continue to spawn goroutines. Cannot pass nil channel. Closing the dagstore will leave these goroutines orphaned.

Alternate design is to return a ticket which the client may optionally use to enquire the status of the shard operation:

```
RegisterShard(key shard.Key, mnt mount.Mount,opts RegisterOpts) (*ShardTicket, error)

```

A ticket is a concrete type:

```
type ShardTicket struct {
	mu       sync.Mutex     
	done     chan struct{}
	result   *ShardResult
}

// Done returns a channel that is closed when the shard operation is complete
func (s *ShardTicket) Done() <-chan struct{} {
	return s.done
}

// Result returns the result of the shard operation or nil if the operation has not completed
func (s *ShardTicket) Result() *ShardResult {
	s.mu.Lock()
	res := s.result
	s.mu.Unlock()
	return res
}

// complete marks the operation as complete
func (s *ShardTicket) complete(result *ShardResult) {
	s.mu.Lock()
	s.result = result
	s.mu.Unlock()
	close(s.done)
}
```

The dagstore simply calls `complete` on the ticket it created when the operation has been completed.

The client can make the dagstore call synchronous:

```
ticket, _ := dagstore.RegisterShard(key, mnt, results, opts)
select {
	case <-ticket.Done():
}
res := ticket.Result()
```

Or asynch, but here the number of goroutines is under the control of the client and can cancelled with to client context:

```
results := make(chan ShardResult, 100)

go func() {
	select {
		case res := <-results:
			// do something with the result
	}
}()


for _, key := range shards {
	ticket, _ := dagstore.RegisterShard(key, mnt, results, opts)
	go func() {
		select {
			case <-ticket.Done():
				results <- ticket.Result()
			case <-ctx.Done():
				return
		}
	}()
}
```

Reply:
@iand The same API usage can be achieved with the current model by passing in a per-request channel. As a user:

```go
ch := make(chan ShardResult, 1)
_ = dagstore.RegisterShard(key, mnt, ch, opts)
<-ch
```

With this model, the client is truly in control of how they want to consume results.

I deliberately don't want the DAG store to return a channel because that's the most inflexible model. If you register many shards at once, you will need to monitor each channel with a separate goroutine (unless you use reflection to select of many channels -- yuck).

Reply:
BTW -- this PR does not guarantee order-safety, which I do want. It's likely that I end up implementing a notifier goroutine on the DAG side whose job is to publish results back to users. So if the user blocks, it won't block the DAG store's event loop, just the result stream.

Reply:
The code you show is identical to the synchronous example I gave except you require the client to know to use a correctly sized channel. 

The main issue I see is the unbounded growth of goroutines in a long running server easily caused by the client holding the API wrong (unbuffered / slow draining / nil channel). The dagstore has no way of detecting when these should be terminated, but the client does.

Reply:
> The code you show is identical to the synchronous example I gave except you require the client to know to use a correctly sized channel.

Client can use any buffer size they want. To keep the system healthy, they should be consuming, but the system won't stop making progress if they fail to do so.

> The main issue I see is the unbounded growth of goroutines in a long running server easily caused by the client holding the API wrong (unbuffered / slow draining / nil channel).

I agree, and as I said above, I plan to move response dispatch to its own, single-threaded goroutine, instead of spinning up ad-hoc goroutines. Doing so will also preserve ordering.

> The dagstore has no way of detecting when these should be terminated, but the client does.

What introduces that ability in your example is the client passing in a `context.Context`, not the DAG store creating and owning a channel. I agree this might be something worth doing, also for the purposes of tracing/metrics, as you also pointed out.

Reply:
> I agree, and as I said above, I plan to move response dispatch to its own, single-threaded goroutine, instead of spinning up ad-hoc goroutines. Doing so will also preserve ordering.

We will have to introduce some sort of backpressure at the point the event loop hands off the result to the result dispatcher go-routine which will block the event loop or start dropping results to the floor.

Reply:
Gotta analyse the tradeoffs carefully.